tests/unit/utils/test_juju_utils.py

```diff
@@ -577,3 +578,83 @@ async def test_disconnect(self):
         mymodel = MagicMock()
         mymodel.disconnect.return_value = "ok"
         await model._disconnect(mymodel)
+
+
+class JujuWaiterTests(aiounittest.AsyncTestCase):
+    def setUp(self):
+        super().setUp()
+
+        self.model_connected = AsyncMock()
+        self.model_connected.info.name = "test"
+        self.model_connected.is_connected = MagicMock()
+        self.model_connected.is_connected.return_value = True
+        self.model_connected.connection = MagicMock()
+        self.model_connected.wait_for_idle = AsyncMock()
+        connection = MagicMock()
+        connection.is_open = True
+        self.model_connected.connection.return_value = connection
+
+        self.model_juju_exception = AsyncMock()
+        self.model_juju_exception.info.name = "test"
+        self.model_juju_exception.is_connected = MagicMock()
+        self.model_juju_exception.is_connected.return_value = True
+        self.model_juju_exception.connection = MagicMock()
+        connection = MagicMock()
+        connection.is_open = True
+        self.model_juju_exception.connection.return_value = connection
+
+    async def raise_juju_unit_exception(**kwargs):
+        raise JujuUnitError()
+
+    async def raise_timeout_exception(**kwargs):
+        raise model.JujuWaiter.TimeoutException()
+
+    async def raise_exception(**kwargs):
+        raise Exception()
+
+    def raise_timeout_exception_sync(*args):
+        raise model.JujuWaiter.TimeoutException()
+
+    def raise_exception_sync(*args):
+        raise Exception()
+
+    async def test_normal(self):
+        waiter = model.JujuWaiter(self.model_connected)
+        await waiter.wait(10)
+
+    async def test_exception(self):
+        self.model_juju_exception.wait_for_idle = JujuWaiterTests.raise_juju_unit_exception
```

Comment:
Why we need this and not `JujuUnitError()` as side_effect?
```suggestion
        self.model_juju_exception.wait_for_idle.side_effect = JujuUnitError()
```